File: DSAlgorithms/Leetcode/c++/binary_search/875-koko_eating_bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
       int left = 1;
       int right = 0;
       for(int i = 0; i < piles.size(); i ++)
            right = std::max(right, piles[i]);
        int result = right;
        while(left <= right)
        {
            int k = left + (right - left)/2;
            long hours = 0;
            for(int i = 0; i < piles.size(); i++)
                hours += ceil((double)piles[i]/k);
            if(hours <= h)
            {
                result = std::min(result, k);
                right = k - 1;
            }
            else
                left = k + 1;
        }
        return result;

    }
};

File: dsa_algo/leetcode/ts/src/binary_search/875-koko_eating_bananas.ts
const minEatingSpeed = (piles: number[], h: number): number =>{
    let left = 1;
    let right = 0;
    for(let i = 0; i < piles.length; i++)
        right = Math.max(piles[i], right);
    let result =  right;
    while(left <= right)
    {
        let k = Math.floor(left + (right - left) /2);
        let hours = 0;
        for(let i = 0; i < piles.length; i++)
            hours += Math.ceil(piles[i]/k);
            if(hours <= h)
            {
                result = Math.min(result, k);
                right = k - 1;
            }
            else
                left = k + 1;
    }
    return result;
};